### PR TITLE
Issue #8018: AST based operatorwrap

### PR DESCRIPTION
**Pull request by pbludov**
Issue #8018

Text-based approach replaced with AST-based. 

Diff Regression projects: https://gist.githubusercontent.com/nmancus1/32ddbff9e69986ac2fcbbeff7def2f80/raw/8480445657f1d215e683dbb6b14c1577bbd44e6b/projects-to-test-on.properties

Diff Regression config: https://gist.githubusercontent.com/pbludov/81b21835641d4f9f61ce70954a29dfcb/raw/a703de479422ac1bcd12e5a69b653217b3513a73/Issue-8018-operator-wrap-config.xml

Diff Regression config (NL): https://gist.githubusercontent.com/pbludov/cb09980e6eef2a89b2298ed10c087fbc/raw/a141f53d13c3907349217758bea066fac3b49c63/Issue-8018-operator-wrap-config-nl.xml

Diff Regression config (EOL): https://gist.githubusercontent.com/pbludov/a233b467c6c3b1ddfd06041fde6577a5/raw/37ba5305aaa37aa1d8623d84804b0a5ed1dad6b0/Issue-8018-operator-wrap-config-eol.xml

### Regression reports
* All tokens, config = Checkstyle: https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/issue-8018-operator-wrap_2020121536/reports/diff/index.html
* All tokens, config = EOL: https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/issue-8018-operator-wrap_2020203715/reports/diff/index.html
* All tokens, config = NL: https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/issue-8018-operator-wrap_2020090445/reports/diff/index.html

### CODE REVIEW

**Comment by pbludov** [comment]
GitHub, generate report

**Comment by pbludov** [comment]
GitHub, generate report

**Comment by pbludov** [comment]
GitHub, generate report

**Comment by pbludov** [comment]
GitHub, generate report

**Comment by pbludov** [comment]
GitHub, generate report

**Comment by pbludov** [comment]
GitHub, generate report

**Comment by pbludov** [comment]
GitHub, generate report

**Comment by pbludov** [comment]
GitHub, generate report

**Review by strkkk — APPROVED** [review]

**Comment by romani** [comment]
Github, rebase

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java [hunk 1]
```diff
@@ -302,28 +304,150 @@ public void setOption(String optionStr) {
...
+     * }
+     *
+     * @param node the node to adjust
+     * @param step the node transformer, should be {@link DetailAST#getPreviousSibling}
+     *             or {@link DetailAST#getNextSibling}
+     * @return adjusted node
+     */
+    private static DetailAST adjustParens(DetailAST node, UnaryOperator<DetailAST> step) {
+        DetailAST result = node;
+        int acc = 0;
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java:437]: what is acc ?
please use full name.

**Review by romani — CHANGES_REQUESTED** [review]
complete rewrite of Check :) 


items:

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java (hunk 1 above)
Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java]: done

**Comment by romani** [comment]
@pbludov , please use `id` for modules next time if you use two instances of the same Check, it will be much easier to read diff report.

**Comment by pbludov** [comment]
> please copy from
done

**Comment by pbludov** [comment]
GitHub, generate report

**Comment by pbludov** [comment]
@romani ping

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java [hunk 2]
```diff
@@ -302,28 +304,150 @@ public void setOption(String optionStr) {
...
+        }
+    }
+
+    /**
+     * Filters some false tokens that this check should ignore.
+     *
+     * @param node the node to check
+     * @return {@code true} for all nodes this check should validate
+     */
+    private static boolean isOperatorNode(DetailAST node) {
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java:326]: description of method is not same a name of method.
should we name method more abstract ... like `isTargetNode`

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java [hunk 3]
```diff
@@ -302,28 +304,150 @@ public void setOption(String optionStr) {
...
-
-            // Check if rest of line is whitespace, and not just the operator
-            // by itself. This last bit is to handle the operator on a line by
-            // itself.
+        if (isOperatorNode(ast)) {
             if (option == WrapOption.NL
-                    && !text.equals(currentLine.trim())
-                    && CommonUtil.isBlank(currentLine.substring(colNo + text.length()))) {
-                log(ast, MSG_LINE_NEW, text);
+                    && TokenUtil.areOnSameLine(ast, getLeftNode(ast))
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java:309]: is it more readable form:
```
if (isOperatorNode(ast)) {
    if (option == WrapOption.NL) {
        if (TokenUtil.areOnSameLine(ast, getLeftNode(ast))
              && !TokenUtil.areOnSameLine(ast, getRightNode(ast))) {
           log(ast, MSG_LINE_NEW, ast.getText());
        }
    }
    else if (option == WrapOption.EOL) {
      if (!TokenUtil.areOnSameLine(ast, getLeftNode(ast))) {
        log(ast, MSG_LINE_PREVIOUS, ast.getText());
      }
    }
}

```

user who debug certain option will not need to real long expressions at all.

**Review by romani — CHANGES_REQUESTED** [review]
items:

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java (hunk 3 above)
Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java:309]: This lead to PMD violation https://pmd.github.io/latest/pmd_rules_vm_design.html#collapsibleifstatements
Maybe it is possible to fool this rule with a switch.

Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java]: Upd: fixed with two additional methods

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java (hunk 2 above)
Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java (hunk 3 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java]: we can suppress Pmd on this case, it is normal that tools have false positives.
we should focus on what is good for reading.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java [hunk 4]
```diff
@@ -302,28 +304,168 @@ public void setOption(String optionStr) {
...
-        final DetailAST parent = ast.getParent();
-        // we do not want to check colon for cases and defaults
-        if (parent.getType() != TokenTypes.LITERAL_DEFAULT
-                && parent.getType() != TokenTypes.LITERAL_CASE) {
-            final String text = ast.getText();
-            final int colNo = ast.getColumnNo();
-            final int lineNo = ast.getLineNo();
-            final String currentLine = getLine(lineNo - 1);
+        if (isTargetNode(ast)) {
+            if (option == WrapOption.NL && isNewLineViolation(ast)) {
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java:308]: please make it like:
`if (option == WrapOption.NL && isNewLineModeViolation(ast)) {`
the same method name update for other mode.

**Review by romani — APPROVED** [review]
ok to merge after last request for method name:

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java (hunk 4 above)
Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java]: done

**Comment by pbludov** [comment]
GitHub, generate report

**Comment by pbludov** [comment]
CI is green

**Pull request merged**